fix: calculate fields for auto email

ludtkemorgan opened the pull request:
# Pull Request Template

## Issue Overview

This PR addresses #368 

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This connects the listing update call to the govDelivery automated email. It also calculates the dynamic information from the listing in order to populate the table information. 

Lastly it wraps the code in a new feature flag tied to the jurisdiction. The migration script adds the new feature flags and translations, but does not turn on the feature.

## How Can This Be Tested/Reviewed?

In order to test this with the govDelivery stage environment you can reach out to @ludtkemorgan to get yourself added to the email subscribe list.

You will also need the following variables setup in your backend .env file. You can get the values for those from @ludtkemorgan as well
```
GOVDELIVERY_USERNAME=
GOVDELIVERY_PASSWORD=
GOVDELIVERY_API_URL=
GOVDELIVERY_TOPIC=
```

The feature flag will also need to be turned on after running the migrations. I'd recommend reseeding and then running the following in order to turn it on:
```
UPDATE jurisdictions
SET enable_listing_opportunity = TRUE
where id = (select id from jurisdictions where name = 'Bay Area')
```

After all of the above is done you can test this by "publishing" any listing on the partner site and then you should receive an email from MTC

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

ColinBuyck commented on backend/core/src/email/email.service.ts:
```diff
@@ -224,34 +255,107 @@ export class EmailService {
     })
 
-    await this.govSend(rawHtml, "Listing Opportunity")
+    await this.govSend(compiled, "Listing Opportunity")
```
Probably should've been handled in my PR but realized that the email subject doesn't match the Figma
<img width="493" alt="Screenshot 2023-11-08 at 11 08 44 AM" src="https://github.com/metrotranscom/doorway/assets/53269332/ff239805-4477-41ef-bcfe-72f084d612cd">

ludtkemorgan replied:
Updated!

ColinBuyck commented on backend/core/src/email/email.service.spec.ts:
```diff
@@ -30,6 +33,18 @@ const application = {
+const httpPostMockFn = jest.fn()
+const mockHttpService = {
+  post: httpPostMockFn.mockImplementation(() => {
+    return of({
```
Do you mind explaining what of() is doing here?

ludtkemorgan replied:
The httpService returns an rxjs observable. This "of" function converts the passed in params to an observable. https://rxjs.dev/api/index/function/of. We don't actually care about the data that is returned, so I could put anything in here. But it needs to be of type observable

ColinBuyck reviewed (changes requested):
Noticing two things: 1) When I published a listing from creation (as in skip draft state), the email doesn't send. 2) Filled out a listing with the bare minimum information and it resulted in NaNs and infinity symbols?. I think we can handle those cases but also curious from a design perspective how empty data should be handled. Should we remove the row all together or leave it empty @eajensenwa?
<img width="604" alt="Screenshot 2023-11-08 at 11 54 48 AM" src="https://github.com/metrotranscom/doorway/assets/53269332/f5b8fbc4-8243-4b09-a65a-9c7b4b975696">

ludtkemorgan commented:
@ColinBuyck thanks for the review! I updated to not show the rows if there is no data for that row. So now rent will only show if at least one unit has a rent set, along with minimum and maximum income. We can still get sign off from design.

Also, I added the flow for creating a listing from scratch and skipping draft state.

ColinBuyck commented on backend/core/src/email/email.service.ts:
```diff
@@ -224,34 +255,115 @@ export class EmailService {
+
+    const compiled = compiledTemplate({
+      listingName: listing.name,
+      listingUrl: `${jurisdiction.publicUrl}/listings/${listing.id}`,
```
I believe this should just be "listing" since currently getting a page not found with listings

ludtkemorgan replied:
Nice catch! I updated this

ColinBuyck reviewed (approved):
LGTM, just one minor change. Also, I'm going to leave this 1 review needed since it's a new feature and would love to get a fresh perspective on something of this scope

YazeedLoonat commented on backend/core/src/email/email.service.ts:
```diff
@@ -20,15 +22,43 @@ import { Jurisdiction } from "../jurisdictions/entities/jurisdiction.entity"
+  label: string,
+  pluralLabel?: string
+): string => {
+  const mappedField = units
```
non-blocking suggestion: changing this to be a `reduce` instead of map and filter would reduce the number of times were iterating through units

YazeedLoonat commented on backend/core/src/email/email.service.ts:
```diff
@@ -224,34 +255,115 @@ export class EmailService {
+        id: listing.jurisdiction.id,
+      },
+    })
+    void (await this.loadTranslations(jurisdiction, Language.en))
```
question: do we need the `void` here?

ludtkemorgan replied:
I don't think it's needed, but it is the pattern in this file so I'm a bit worried to pull it out for this for some weird edge case. When we migrate to prisma we won't use this pattern so I'd vote to get rid of the void at that point

YazeedLoonat commented on backend/core/src/email/email.service.ts:
```diff
@@ -224,34 +255,115 @@ export class EmailService {
     if (this.configService.get<string>("NODE_ENV") == "production") {
       Logger.log(
-        `Preparing to send a listing opportunity email to ${user.email} from ${jurisdiction.emailFromAddress}...`
+        `Preparing to send a listing opportunity email for ${listing.name} from ${jurisdiction.emailFromAddress}...`
```
suggestion: maybe dropping the trailing `...`

ludtkemorgan replied:
This is the pattern in the other loggers here. However, I don't think I have ever seen these logs in the logs before.

YazeedLoonat reviewed (approved):
looks good! just a few questions
